packages/wranglerjs-compat-webpack-plugin/src/__tests__/helpers/compare-outputs.ts

```diff
@@ -0,0 +1,249 @@
+        console.log(output);
+        console.warn(`WARNING ${warning}`);
+      }
+    }),
+  });
+  const stderr = new WritableStream({
+    write: pipe(console.error),
+  });
+
+  try {
+    await withCapturedChildProcessOutput(
+      async () => await runWrangler2("publish"),
+      {
+        stdout,
+        stderr,
+      }
+    );
+  } catch (e) {
+    wrangler2result = e as Error;
+  } finally {
+    process.stdout.unpipe(stdout);
+    process.stderr.unpipe(stderr);
+  }
+
+  // an assertion failed, so we should throw
+  if (wrangler2result !== undefined && isAssertionError(wrangler2result)) {
+    throw wrangler2result;
+  }
+
+  const wrangler2 = {
+    result: wrangler2result,
+    std: {
+      out: std.out,
+      err: std.err,
+      warn: std.warn,
+    },
+    output: walk(path.join(wrangler2Dir, "worker")),
+  };
+
+  return { wrangler1, wrangler2 };
+}
+
+/**
+ * Clear the console by resetting mocks to console.log, .error, and .warn
+ */
+const clearConsole = () => {
+  (console.log as jest.Mock).mockClear();
+  (console.warn as jest.Mock).mockClear();
+  (console.error as jest.Mock).mockClear();
+};
+
+/**
+ * Jest errors aren't exported directly as a type, so this hacky garbage
+ * checks if an error has a "matcherResult" property, which all jest errors
+ * have.
+ */
+const isAssertionError = (e: Error) =>
+  Object.prototype.hasOwnProperty.bind(e)("matcherResult");
+
+/**
+ * Temporarily capture output of processes spawned by `child_process.spawn()`
+ * when the `fn` arg is called. All output will be piped through to the
+ * `stdout` and `stderr` args, which can be arbitrary streams.
+ *
+ * Useful for capturing output from webpack when called by wrangler 2.
+ * @returns the result of calling `fn`
+ */
+async function withCapturedChildProcessOutput<T>(
+  fn: () => T | Promise<T>,
+  { stdout, stderr }: { stdout: WritableStream; stderr: WritableStream }
+): Promise<T> {
+  const { spawn } = childProcess;
+  let process: childProcess.ChildProcess | undefined = undefined;
+  const childProcessMock = jest
+    .spyOn(childProcess, "spawn")
+    .mockImplementation((command, args, options) => {
+      process = spawn(command, args, options);
+      if (process.stdout !== null && process.stderr !== null) {
+        process.stdout.pipe(stdout);
+        process.stderr.pipe(stderr);
+      }
+      return process;
+    });
+
+  try {
+    return await fn();
+  } finally {
+    if (process.stdout !== null && process.stderr !== null) {
+      process.stdout.unpipe(stdout);
+      process.stderr.unpipe(stderr);
+    }
+    childProcessMock.mockRestore();
+  }
+}
+
+/**
+ * Walk a directory, reading all files into an object keyed by their filenames
+ */
+function walk(dir: string): DirectoryContent {
+  const entries: DirectoryContent = {};
+
+  fs.readdirSync(dir).forEach((entry) => {
+    const entryPath = path.resolve(dir, entry);
+    if (fs.lstatSync(entryPath).isDirectory()) {
+      entries[entry] = walk(entryPath);
```

Comment:
won't files with same names in different dirs clash here? should this be `dir + entry`? (I'm not actually sure, hence asking)

Reply:
ummm...i don't think so? Because they'll be like
```
- parent
  | - child
  | - child2
    | - child
```
should become
```
{
  child: <Buffer>
  child2: {
    child: <Buffer>
  }
}
```
right?